src/devnet/integrated-devnet.ts

```diff
@@ -27,28 +29,61 @@ export abstract class IntegratedDevnet {
     protected abstract cleanup(): void;
 
     public async start(): Promise<void> {
+        // Do not accept stale existing devnets, because they might
+        // be running an older incompatible version of the server
+        if (await this.isServerAlive()) {
+            throw new Error(`integrated-devnet cannot start because existing devnet already running on ${this.host}:${this.port}`)
+        }
+
         this.childProcess = await this.spawnChildProcess();
 
         return new Promise((resolve, reject) => {
+            // called on successful start of the child process
             this.childProcess.on("spawn", async () => {
+                const startTime = new Date().getTime();
                 const maxWaitMillis = 60_000;
-                const oneSleepMillis = 500;
-                const maxIterations = maxWaitMillis / oneSleepMillis;
-                for (let i = 0; i < maxIterations; ++i) {
-                    await sleep(oneSleepMillis);
-                    try {
-                        await axios.get(`http://${this.host}:${this.port}/${DEVNET_ALIVE_URL}`);
+                const oneSleepMillis = 100;
+
+                // keep checking until process has failed/exited
+                while (this.childProcess) {
+                    const elapsedMillis = new Date().getTime() - startTime;
+                    if (elapsedMillis >= maxWaitMillis) {
+                        reject(new Error(`integrated-devnet connection timed out!`));
+                        break;
+                    }
+                    if (await this.isServerAlive()) {
+                        this.connected = true;
                         resolve();
-                    } catch (err: unknown) {
-                        // cannot connect yet, devnet is not up
+                        break;
+                    } else {
+                        await sleep(oneSleepMillis);
                     }
                 }
-                reject(`Could not connect to integrated-devnet in ${maxWaitMillis} ms!`);
             });
-
+            
+            // this only happens if childProcess completely fails to start
             this.childProcess.on("error", (error) => {
+                this.childProcess = null;
                 reject(error);
             });
+
+            // capture the most recent message from stderr
+            this.childProcess.stderr.on("data", (data: any) => {
```

Comment:
This is up to how the internal application flushes its stderr. It's out of our control. For most use cases this will work as expected, eg: `print('something went wrong!', file=sys.stderr, flush=True)` will give `something went wrong!\n`